Rework function spaces

wence- opened the pull request:
This completely changes the way function space caching is done, to make
it more robust, share more data, and remove some of the magic.

We now build up function spaces in the following manner.

1. Given fiat entity_dofs, we build the shared data (this contains the
   PyOP2 Sets and Maps).  This data is cached on the mesh of the
   function space.  Because tensor-valued and their respective
   scalar-valued elements have the same entity dofs, this means we now
   share heavy data between these spaces.

2. All function space types are now either FunctionSpace or
   MixedFunctionSpace.  The distinguishing factor between vector/tensor
   and so forth is their rank.

3. FunctionSpaces are lightweight objects (the shared data is heavy and
   compared with object identity)

4. As a consequence we can get rid of the mess of IndexedFunctionSpace
   and IndexedVFS and replace them with a ProxyFunctionSpace,
   essentially just a normal FunctionSpace on which we can attach
   additional properties.  This signficantly simplifies things.

Finally, the user interface to constructing FunctionSpaces (and
MixedFunctionSpaces) becomes completely functional.  As a result,
there's no need to do horrible dances with __new__ when building the
FunctionSpace classes (the user just calls a function which returns the
correct object instance).

Finally, we come up to date with FEniCS compatibility by supporting
FunctionSpace(mesh, some_finite_element) to build scalar-, vector-,
tensor- and mixed function spaces.

miklos1 commented:
I made a sketch on how to make caching even more precise, so that every piece of data depends only on the key that we need to calculate it: https://gist.github.com/miklos1/a7532d9ae4e1bcf33e4b

I don't find the more precise caching necessary for the merge, but it's nice to have the idea linked here.

wence- commented:
Thanks.

After discussion at lunchtime, I have also re-allowed the ability to say `VectorFunctionSpace(mesh "RT", 1)` to get a tensor-valued space that stores intrinisically vector-valued dofs at each node.  The same is not allowed for `TensorFunctionSpace` since at the moment, UFL requires empty `value_shape` for `TensorElement`.

wence- commented on firedrake/functionspaceimpl.py:
```diff
@@ -0,0 +1,736 @@
+                self.dof_dset is other.dof_dset and \
+                self.ufl_element() == other.ufl_element()
+        except AttributeError:
+            return False
```
Is this the correct equality comparison?  I think if the meshes match, and the ufl element matches, the dof_dset must also match, but I was belt-and-braces here.

wence- replied:
Note that this relies on the `ObjectCached` inheritence of `DataSet` to match up `DataSet` instances on the same `Set` with the same `shape`.

wence- commented on firedrake/functionspaceimpl.py:
```diff
@@ -0,0 +1,736 @@
+        # to the function space) could be shared between
+        # functionspaces with the same dof_dset.
+        with self.make_dat().vec_ro as v:
+            dm.setGlobalVector(v.duplicate())
```
I claim `FunctionSpaces` are now lightweight, however this is slightly bad.  I need one dm per `FunctionSpace` instance (that's OK, it's not big).  The `pointSF` is shared and the default section is shared.  However, the `Vec` is not shared.  So each instance of a `FunctionSpace` that accesses the `_dm` property makes a new `Vec` (i.e. a new `Dat`).  The Vec can be shared with everything that has the same number of `nodes_per_entity` which gives the global numbering, and also the same dofs per node (i.e. the function space shape).  Miklos' suggestion of caching individual bits would fix this problem I think.

wence- commented on firedrake/functionspaceimpl.py:
```diff
@@ -0,0 +1,736 @@
+    ``None``."""
+
+    def __eq__(self, other):
+        try:
```
Fast path for `type(other) is not FunctionSpace`?

miklos1 replied:
I think it is in general a good idea to check types first in equality predicates.

wence- commented on firedrake/functionspaceimpl.py:
```diff
@@ -0,0 +1,736 @@
+        # FIXME: The Vec could be shared between functionspaces with
+        # the same dof_dset.
+        with self.make_dat().vec_ro as v:
+            dm.setGlobalVector(v.duplicate())
```
As before, this is wasteful (one Vec per MFS instance)

wence- commented on firedrake/functionspaceimpl.py:
```diff
@@ -0,0 +1,736 @@
+        subdm = subspace._dm
+        # Keep hold of strong reference to created subspace (given we
+        # only hold a weakref in the shell DM)
+        W._subspaces.append(subspace)
```
Return same subspace if asked for this again?  Now that `MFS` is not object cached we probably should.

wence- commented on firedrake/functionspaceimpl.py:
```diff
@@ -0,0 +1,736 @@
+        W._subspaces.append(subspace)
+        # Index set mapping from W into subspace.
+        iset = PETSc.IS().createGeneral(numpy.concatenate([W._ises[f].indices
+                                                           for f in fields]))
```
Same for the `IS`.  Note that these will be recreated at each iteration of a nonlinear solver if complicated fieldsplits are applied.  The subspaces info could be keyed by the fields passed in.

wence- commented on firedrake/functionspaceimpl.py:
```diff
@@ -0,0 +1,736 @@
+        if mesh is not topology:
+            return WithGeometry(self, mesh)
+        else:
+            return self
```
Could remove need for new by making a `ProxyFunctionSpaceBase` and making `ProxyFunctionSpace` a function not a class.  But it's not too complicated, so I don't think there's a need.  Note that argument sanitisation doesn't take place here.  The use case is, there is an existing `FunctionSpace` and now we need to add some properties.

wence- replied:
Should this constructor therefore take an existing functionspace (rather than a mesh and element).

wence- commented on firedrake/mesh.py:
```diff
@@ -976,7 +975,7 @@ def make_mesh_from_coordinates(coordinates):
 
     coordinates_fs = coordinates.function_space()
-    if not isinstance(coordinates_fs, functionspace.VectorFunctionSpace):
+    if coordinates_fs.rank != 1:
```
Should one additionally assert that the `value_shape` has length 1?  The previous check did not do this.

miklos1 replied:
Yes!

wence- commented:
I've implemented Miklos' suggestion of more precise caching and also fixed up an inefficiency in the way splits were created for nested fieldsplits of mixed function spaces.  I claim this is now good to go.  Comments still welcome though.

miklos1 commented on firedrake/functionspacedata.py:
```diff
@@ -0,0 +1,517 @@
+
+    :arg mesh: The mesh to use.
+    :arg entity_dofs: The FIAT entity_dofs dict.
+    :arg global_numbering: The PETSc Section describing node layout
```
So `global_numbering` is cached on a key (`nodes_per_entity`) that is less distinguishing than `entity_dofs`, therefore it is fine.

miklos1 replied:
Oh, wait, this function is not `@cached`.

wence- replied:
Yes, but the same argument applies to get_entity_node_lists below, which is cached.

miklos1 commented:
I think it looks good.
